src/python/pants/engine/rules.py

```diff
@@ -191,13 +192,24 @@ def rule_decorator(func, **kwargs) -> Callable:
         name=f"{func_id} return",
         raise_type=MissingReturnTypeAnnotation,
     )
+
+    func_params = inspect.signature(func).parameters
+    for parameter in param_type_overrides:
+        if parameter not in func_params:
+            raise ValueError(
+                f"Unknown parameter name in `param_type_overrides`: {parameter}."
+                + f" Parameter names: '{', '.join(func_params)}'"
+            )
+
     parameter_types = tuple(
-        _ensure_type_annotation(
+        param_type_overrides[parameter]
+        if parameter in param_type_overrides
```

Comment:
I think this bypasses the type check performed by `_ensure_type_annotation`, maybe check that in the previous for loop?
```python
for parameter, value in param_type_overrides.items():
  ...
```